File: src/Threading/Python/Producer_Consumer.py
import threading
import time
import random

product = 0
condition = threading.Lock() # for critical section
quanta = 1

class Producer(threading.Thread):

    def __init__(self, val='Thread'):
        self._name = val
    
    @property
    def name(self):
        return self._name
    
    @name.setter
    def name(self, val):
        self._name = str(val)
    
    def run(self):
        global product
        count = time.time()
        while time.time()-count < quanta:
            f = random.randrange(1, 5)
            condition.acquire()
            product = product + f
            time.sleep(random.random()/10)
            print(f"{self.name} : Add-{f}     | Stock-{product}")
            condition.release()

class Consumer(threading.Thread):

    def __init__(self, val='Thread'):
        self._name = val
    
    @property
    def name(self):
        return self._name
    
    @name.setter
    def name(self, val):
        self._name = str(val)

    def run(self):
        global product
        count = time.time()
        while time.time()-count < quanta:
            f = random.randrange(5, 10)
            condition.acquire()
            if f >= product:
                print(f"{self.name} : Require-{f} | Stock-{product} | Get-{product} | Starve-{f-product}")
                product = 0
            else:
                print(f"{self.name} : Require-{f} | Stock-{product} | Get-{f} | Starve-0")
                product = product - f
            time.sleep(random.random()/10)
            condition.release()
                
def main():
    global product
    product = 0
    print(f"Product : {product}")
    producer_ref = Producer('P')
    consumer_ref = Consumer('C')
    producer_thread = threading.Thread(target=producer_ref.run, args=())
    consumer_thread = threading.Thread(target=consumer_ref.run, args=())

    producer_thread.start()
    consumer_thread.start()

    producer_thread.join()
    consumer_thread.join()

    print(f"Product : {product}")

main()

File: src/Threading/Java/Producer_Consumer.java
import java.lang.Math;

class Product{
    private int product = 0;

    public int get_product(){
        return this.product;
    }

    public synchronized void get_product(String name, int query){
        int val = 0, starve = 0, stock = this.product;
        if(this.product <= query){
            val = this.product;
            starve = query-this.product;
            this.product = 0;
        }else{
            val = query;
            this.product = this.product-query;
        }
        System.out.println(name+" : Require-"+query+" | Stock-"+stock+" | Get-"+val+" | Starve-"+starve);
    }

    public synchronized void put_product(String name, int query){
        this.product = this.product + query;
        System.out.println(name+" : Add-"+query+" | Stock-"+this.product);
    }
}

class Producer extends Thread{
    
    private String name;
    private Product p_ref;

    public Producer(Product p_ref, String name){
        this.p_ref = p_ref;
        this.name = name;
    }

    public void run(){
        int i, f;
        for(i=0; i<10; i++){
            f = (int)(Math.random()*10) + 1;
            this.p_ref.put_product(this.name,f);
        }
    }
}

class Consumer extends Thread{
    
    private String name ;
    private Product p_ref;

    public Consumer(Product p_ref, String name){
        this.p_ref = p_ref;
        this.name = name;
    }

    public void run(){
        int i, f;
        for(i=0; i<10; i++){
            f = (int)(Math.random()*10) + 5;
            this.p_ref.get_product(this.name,f);
        }
    }
}

public class Producer_Consumer{

    public static void main(String[] args)throws Exception{
        Product product_ref = new Product();
        Producer producer_ref = new Producer(product_ref, "Producer");
        Consumer consumer_ref = new Consumer(product_ref, "Consumer");

        System.out.println("Products : "+product_ref.get_product());

        producer_ref.start();
        consumer_ref.start();
        /*
        In case of implements Runnable;
        Eg:
        Producer p_ref = new Producer(...);  implements Runnable
        Thread t = new Thread(p_ref); other wise it would be a part of main thread
        t.start();
        */
        producer_ref.join();
        consumer_ref.join();

        System.out.println("Products : "+product_ref.get_product());
    }
}